## Analytic Decision Making Homework 1

### Exercise 1: Operations Excellence

#### 1. Write the optimization models for each plant

**BIC**:

**Decision Variables**: $x_1$ (standard units), $x_2$ (deluxe)

**Objective Function**: maximize $10x_1 + 15x_2$

**Contraints**:

$ 4x_1 + 2x_2 \le 80$

$ 2x_1 + 5x_2 \le 60$

$ 4x_1 + 4x_2 \le 75$

$ x_1, x_2 \ge 0$



**CIB**:

**Decision Variables**: $y_1$ (standard units), $y_2$ (deluxe)

**Objective Function**: maximize $10y_1 + 15y_2$

**Contraints**:

$ 5y_1 + 3y_2 \le 60$

$ 5y_1 + 6y_2 \le 75$

$ 4y_1 + 4y_2 \le 45$

$ y_1, y_2 \ge 0$

#### 2. Provide the optimal solutions and profit levels for each factory. Give the values of the decision variables and the optimal objective function value.

In [ ]:
# this section of code contains partial help from Claude and previous class exercises.
import cvxpy as cvx

# Define decision variables
x1 = cvx.Variable(integer=True)  # Standard product
x2 = cvx.Variable(integer=True)  # Deluxe product

y1 = cvx.Variable(integer=True)  # Standard product
y2 = cvx.Variable(integer=True)  # Deluxe product

# Constraints
con_x = [
    4 * x1 + 2 * x2 <= 80,  # Cooking capacity
    2 * x1 + 5 * x2 <= 60,  # Filtering capacity
    4 * x1 + 4 * x2 <= 75,  # Raw material
    x1 >= 0, x2 >= 0
]

con_y = [
    5 * y1 + 3 * y2 <= 60,  # Cooking capacity
    5 * y1 + 6 * y2 <= 75,  # Filtering capacity
    4 * y1 + 4 * y2 <= 45,  # Raw material
    y1 >= 0, y2 >= 0
]

# Objective function
obj_x = cvx.Maximize(10 * x1 + 15 * x2)
obj_y = cvx.Maximize(10 * y1 + 15 * y2)

# Solve the problem
problem_x = cvx.Problem(obj_x, con_x)
problem_y = cvx.Problem(obj_y, con_y)

problem_x.solve()
problem_y.solve()

# Print the optimal solution
print("Optimal solution for BIC Site:")
print(f"Standard product (x1): {x1.value}")
print(f"Deluxe product (x2): {x2.value}")
print(f"Maximum profit: {round(problem_x.value,2)}")

print("\nOptimal solution for CIB Site:")
print(f"Standard product (y1): {y1.value}")
print(f"Deluxe product (y2): {y2.value}")
print(f"Maximum profit: {round(problem_y.value,2)}")

Optimal solution for BIC Site:
Standard product (x1): 10.0
Deluxe product (x2): 8.0
Maximum profit: 220.0

Optimal solution for CIB Site:
Standard product (y1): 0.0
Deluxe product (y2): 11.0
Maximum profit: 165.0


#### 3. Write the optimization model for the joint optimization of both sites.

In [ ]:
# this section of code contains partial help from Claude and previous class exercises.
import cvxpy as cvx

# Define decision variables
x1 = cvx.Variable(integer=True)  # Standard product at BIC
x2 = cvx.Variable(integer=True)  # Deluxe product at BIC
y1 = cvx.Variable(integer=True)  # Standard product at CIB
y2 = cvx.Variable(integer=True)  # Deluxe product at CIB

# Objective function
profit = 10 * x1 + 15 * x2 + 10 * y1 + 15 * y2

# Constraints
constraints = [
    4 * x1 + 2 * x2 <= 80,  # Cooking capacity for BIC
    2 * x1 + 5 * x2 <= 60,  # Filtering capacity for BIC
    5 * y1 + 3 * y2 <= 60,  # Cooking capacity for CIB
    5 * y1 + 6 * y2 <= 75,  # Filtering capacity for CIB
    4 * x1 + 4 * x2 + 4 * y1 + 4 * y2 <= 120,  # Raw material constraint
    x1 >= 0, x2 >= 0, y1 >= 0, y2 >= 0
]

# Problem formulation
problem = cvx.Problem(cvx.Maximize(profit), constraints)

# Solve the problem
problem.solve()

# Print the optimal solution
print("Optimal solution for joint optimization:")
print(f"Standard product at BIC (x1): {x1.value}")
print(f"Deluxe product at BIC (x2): {x2.value}")
print(f"Standard product at CIB (y1): {y1.value}")
print(f"Deluxe product at CIB (y2): {y2.value}")
print(f"Maximum profit: {round(problem.value,2)}")

Optimal solution for joint optimization:
Standard product at BIC (x1): 10.0
Deluxe product at BIC (x2): 8.0
Standard product at CIB (y1): 0.0
Deluxe product at CIB (y2): 12.0
Maximum profit: 400.0


#### 4. Please comment on the results

After joining the 2 sites, we do not need to worry about the distribution of raw materials. We can now optimize the use of the entire 120kg of raw materials without reaching a boundary on other contraints. Thus, leading to a better overall profit.

#### 5. Recommendations to the CEO to improve the firm's operations.

Here are some potential recommendations:

1. From running the joint optimization, we notice that the allocation of raw material is not optimal, causing the profit not to be maximized. Please considering reallocating the raw materials between the 2 sites in order to better balance production and maximize overall profit.
2. From another perspective, the production capacity at BIC is not enought to handle the amount of current raw material level. Therefore, considering expansion or upgrades to increase production capabilities is another appraoch to maximize profit.

### Exercise 2: Hotel L'Escargot

#### 1. According to the regression equation, which variables positively/negatively affect Profitability?

"Price of the inn" and "college students in the Area" are 2 variables that positively impact profitability. On the other hand, "state population per inn" and "square root of the median income in the area" negatively affect profitability. Intuitively, this does not make sense because the higher the median income, the more money the population tend to spend, thus creating more potential profit for the inn.

#### 2. Using the spreadsheet, compute the profitability for each hotel. Which one has the highest profitabiity? Which one has the lowest profitability?

In [ ]:
# this section of code contains partial help from Claude
import pandas as pd

# Load the Excel file
xls = pd.ExcelFile('Lescargot.xlsx')
data = pd.read_excel(xls, sheet_name='Sheet1', header=2)

# Write a function to calculate profitability
def compute_profitability(row):
    state_pop_per_inn = row['State Population Per Inn (normalized)']
    price = row['Price (normalized)']
    sqrt_median_income = row['Square Root of Median Income (normalized)']
    college_students = row['College Students in Area (normalized)']

    profitability = 39.05 - (5.41 * state_pop_per_inn) + (5.86 * price) - (3.09 * sqrt_median_income) + (1.75 * college_students)
    return profitability

# Apply the function to each row of the DataFrame
data['Profitability'] = data.apply(compute_profitability, axis=1)

# Print the DataFrame with the computed profitability
print(data.sort_values(by='Profitability', ascending=False))

    Hotel                      Location     Price  Price (normalized)  \
1       2            Fresno, California  10000000            1.699076   
5       6        Long Beach, California   8950000            1.402123   
0       1            Eureka, California   2925000           -0.301823   
2       3            Fresno, California   3750000           -0.068503   
3       4            Fresno, California   3500000           -0.139206   
13     14  South Lake Tahoe, California   3750000           -0.068503   
11     12  South Lake Tahoe, California   2500000           -0.422019   
12     13  South Lake Tahoe, California   1975000           -0.570495   
9      10  South Lake Tahoe, California   1650000           -0.662409   
14     15  South Lake Tahoe, California   1475000           -0.711902   
10     11  South Lake Tahoe, California   1125000           -0.810886   
15     16  South Lake Tahoe, California    750000           -0.916941   
4       5            Fresno, California    325000  

/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


According to the regression equation, hotel 2 has the highest profitability, and hotel 8 has the lowest.

#### 3. L'Escargot has a budget of \$10 million to spend on hotels. Suppose we use a “greedy” approach where we select the most profitable hotels until we ran out of budget. So we would start by the most profitable, and then if we had enough budget left, we would buy the hotel we predict to be the second most profitable, and so on.
* **Describe what we would do with this approach, i.e., which hotels would we purchase?**

  Using this method, we can only afford hotel 2.

* **What would our total predicted profitability be? (This is the sum of the predicted profitability of all hotels we purchase.)**

  The total profitability is 53.38

* **If we are trying to maximize our total predicted profitability, is this a good approach? How about if we were trying to maximize the average predicted profitability of the hotels we select? How about if we had a budget of \$20 million instead of \$10 million?**

  No matter if we are trying to maximize our total predicted profitability, average profitability, or if we have higher budget, this is not the best approach. Instead, we could use linear programming to find the optimal way of buying hotels.

#### 4. Now, build an optimization model to select hotels given the $10 million budget.




**Write out the optimization problem. Make sure to detail the decision variables, the objective function and the constraints.**

* *Decision variable* :
  
  $x_i$ (binary variable indicating whether a hotel is chosen or not)

* *Objective function* :

  maximize $\sum\limits_{i=1}^n (profitability_i × x_i)$


* *Constraints* :

  $\sum\limits_{i=1}^n (price_i × x_i) \le 10,000,000$

In [ ]:
#ignore warning messages
import warnings
warnings.filterwarnings("ignore")

In [ ]:
# this section of code contains partial help from Claude and previous class exercises.
import cvxpy as cp
import pandas as pd

# Define decision variables
num_hotels = len(data)
x = cp.Variable(num_hotels, boolean=True) # binary

# Objective function: Maximize total predicted profitability
profitability = data['Profitability'].values
objective = cp.Maximize(profitability @ x)

# Budget constraint
prices = data['Price'].values
budget = 10000000
budget_constraint = cp.sum(prices * x) <= budget

# Problem formulation
problem = cp.Problem(objective, [budget_constraint])

# Solve the problem
problem.solve()

# Print the selected hotels and their total predicted profitability
selected_hotels = data.iloc[x.value.astype(bool)]
total_predicted_profitability = sum(selected_hotels['Profitability'])

print(x.value)
print("Selected hotels:")
print(selected_hotels[['Hotel','Price','Profitability']])
print(f"Total price: ${sum(selected_hotels['Price'])}")
print(f"Total predicted profitability: {round(total_predicted_profitability,2)}")

[0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 1. 1.]
Selected hotels:
    Hotel    Price  Profitability
4       5   325000      37.344988
9      10  1650000      38.880673
10     11  1125000      38.010600
11     12  2500000      40.289363
12     13  1975000      39.419290
14     15  1475000      38.590649
15     16   750000      37.389119
Total price: $9800000
Total predicted profitability: 269.92


**What is the optimal solution? Give the values of the decision variables and the optimal objective function value.**

Here, we chose hotels 5, 10, 11, 12, 13, 15, and 16. With the total price at \$9,800,000, and predicted profitability of 269.92.


**Does the optimal solution make sense intuitively? How does it compared to the greedy solution?**

The optimal solution makes sense intuitively, where it chooses more hotels with a lower price but competitive profitabilities. Even though this solution did not used up all the budget, it has way higher profitability compared to the greedy solution.

#### 5. L'Escargot thinks that buying too many hotels in one city is probably not a good idea and would prefer to diversify across as many cities as possible. Add constraint(s) to your model to limit the number of hotels purchased in any city to at most 2.

**What are the constraints that you need to add to the model? Intuitively, do you expect the new optimal objective function value to be larger, smaller or the same as before?**

The new constraints we need to add:

$\sum\limits_{i=1}^5 x_i ≤ 2$

$\sum\limits_{i=6}^9 x_i ≤ 2$

$\sum\limits_{i=10}^16 x_i ≤ 2$

Intuitively, I expect the optimal objective function value to be smaller than before since more constraints are added to it.


**Write the new optimization model.**
* *Decision variable* :
  
  $x_i$ (binary variable indicating whether a hotel is chosen or not)

* *Objective function* :

  maximize $\sum\limits_{i=1}^n (profitability_i × x_i)$


* *Constraints* :

  $\sum\limits_{i=1}^n (price_i × x_i) \le 10,000,000$
  
  $\sum\limits_{i=1}^5 x_i ≤ 2$

  $\sum\limits_{i=6}^9 x_i ≤ 2$

  $\sum\limits_{i=10}^16 x_i ≤ 2$



**Solve the new model. Give the values of the decision variables and the optimal objective function value. How does this compare to the previous solution?**

In [ ]:
# this section of code contains partial help from Claude and previous class exercises.
import cvxpy as cp
import pandas as pd

# Define decision variables
num_hotels = len(data)
x = cp.Variable(num_hotels, boolean=True) # binary

# Objective function: Maximize total predicted profitability
profitability = data['Profitability'].values
objective = cp.Maximize(profitability @ x)

# Budget constraint
prices = data['Price'].values
budget = 10000000
budget_constraint = cp.sum(prices * x) <= budget

# new constraints
cons = [
    x[0]+x[1]+x[2]+x[3]+x[4] <= 2,
    x[5]+x[6]+x[7]+x[8] <= 2,
    x[9]+x[10]+x[11]+x[12]+x[13]+x[14]+x[15] <= 2
]

constraints = [budget_constraint] + cons

# Problem formulation
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()

# Print the selected hotels and their total predicted profitability
selected_hotels = data.iloc[x.value.astype(bool)]
total_predicted_profitability = sum(selected_hotels['Profitability'])

print(x.value)
print("Selected hotels:")
print(selected_hotels[['Hotel','Price','Profitability']])
print(f"Total price: ${sum(selected_hotels['Price'])}")
print(f"Total predicted profitability: {round(total_predicted_profitability,2)}")

[1. 0. 0. 0. 1. 0. 1. 1. 0. 1. 1. 0. 0. 0. 0. 0.]
Selected hotels:
    Hotel    Price  Profitability
0       1  2925000      44.242369
4       5   325000      37.344988
6       7  1950000      23.776866
7       8  1750000      23.445409
9      10  1650000      38.880673
10     11  1125000      38.010600
Total price: $9725000
Total predicted profitability: 205.7


The optimal objective function value is 205.7, which is a lot lesser than the previous value.

#### 6. Describe how you would present your results to L'Escargot. Do you have any recommendations for them to improve the regression model? How about to improve the optimization model?

Even though the linear regression model has all variables significant, there could still be potential bias and interaction terms that are not considered. I recommend to go further than simply a linear regression model. Instead, the trend could be non-linear and even time related.

In terms of optimization model, it is all built based on the foundation of the linear regression model. In this case, it is better to not limit 2 number of hotels in any city, so that total profitability can be optimized.

### Exercise 3: Matching for a Dating Platform

#### 1. Assume all of the individuals in the data set would like to be matched to people of the opposite gender. What matches would you choose to maximize the aggregate match score?

In [ ]:
import pandas as pd
import numpy as np
from numpy import *
import math
from math import *
import cvxpy as cvx
from cvxpy import *

data_url='https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv'
df = pd.read_csv(data_url) #read the dataset

k=1
nrow=df.shape[0] #yields the number of rows
ncol=df.shape[1] #yields the number of columns

female=matrix(df[:int(round(nrow/2))].iloc[:,2:].values) #take the numerical values for women
male=matrix(df[int(round(nrow/2)):].iloc[:,2:].values) #take the numerical values for men

score=matrix(male.dot(female.transpose()).transpose(),'d')

x=cvx.Variable((int(nrow/2),int(nrow/2)),boolean=True)

objective = cvx.Maximize(sum(cvx.multiply(score,x)))

c0=(sum(x,axis=0))<=k
c1=(sum(x,axis=1))<=k
con=[c0,c1]

prob = cvx.Problem(objective, con)
result = prob.solve()
print(x.value)

[[0. 0. 0. 0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0.]]


In [ ]:
# this section of code contains partial help from Claude
# Create a list of matches
matches = []
for i in range(int(nrow/2)):
    for j in range(int(nrow/2)):
        if x.value[i, j] == 1:
            female_index = i
            male_index = j + int(round(nrow/2))
            male_name = name_mapping[male_index]
            female_name = name_mapping[female_index]
            match_score = score[i, j]
            matches.append({'Female': female_name, 'Male': male_name, 'Match Score': match_score})

# Convert the matches list to a DataFrame
matches_df = pd.DataFrame(matches)

# Print the matches in a table format
print("Optimal matches to maximize aggregate match score:")
print(matches_df)

Optimal matches to maximize aggregate match score:
     Female    Male  Match Score
0     Laura    Carl          8.0
1  Jennifer    Adam          6.0
2      Emma  Daniel          8.0
3    Olivia   Peter          4.0
4    Martha    Eric          4.0
5  Isabella   Ralph          8.0
6      Mary   James          5.0
7    Sophia  Robert          5.0


#### 2. How would the results change if each person is presented with two possible matches?

In [ ]:
data_url='https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv'
df = pd.read_csv(data_url) #read the dataset

k=2 # allow 2 matches
nrow=df.shape[0] #yields the number of rows
ncol=df.shape[1] #yields the number of columns

female=matrix(df[:int(round(nrow/2))].iloc[:,2:].values) #take the numerical values for women
male=matrix(df[int(round(nrow/2)):].iloc[:,2:].values) #take the numerical values for men

score=matrix(male.dot(female.transpose()).transpose(),'d')

x=cvx.Variable((int(nrow/2),int(nrow/2)),boolean=True)

objective = cvx.Maximize(sum(cvx.multiply(score,x)))

c0=(sum(x,axis=0))<=k
c1=(sum(x,axis=1))<=k
con=[c0,c1]

prob = cvx.Problem(objective, con)
result = prob.solve()
print(x.value)

[[0. 0. 0. 0. 0. 1. 1. 0.]
 [0. 1. 0. 0. 0. 1. 0. 0.]
 [1. 0. 0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 1. 1.]
 [0. 0. 0. 1. 0. 0. 0. 0.]
 [0. 0. 1. 0. 0. 0. 0. 1.]
 [1. 0. 1. 0. 0. 0. 0. 0.]
 [0. 1. 0. 1. 0. 0. 0. 0.]]


In [ ]:
# this section of code contains partial help from Claude
# Create a list of matches
matches = []
for i in range(int(nrow/2)):
    for j in range(int(nrow/2)):
        if x.value[i, j] == 1:
            female_index = i
            male_index = j + int(round(nrow/2))
            male_name = name_mapping[male_index]
            female_name = name_mapping[female_index]
            match_score = score[i, j]
            matches.append({'Female': female_name, 'Male': male_name, 'Match Score': match_score})

# Convert the matches list to a DataFrame
matches_df = pd.DataFrame(matches)

# Print the matches in a table format
print("Optimal matches to maximize aggregate match score:")
print(matches_df)

Optimal matches to maximize aggregate match score:
      Female    Male  Match Score
0      Laura    Adam          6.0
1      Laura    Carl          8.0
2   Jennifer  Robert          2.0
3   Jennifer    Adam          6.0
4       Emma   James          6.0
5       Emma  Daniel          8.0
6     Olivia    Carl          7.0
7     Olivia   Ralph          5.0
8     Martha    Eric          4.0
9   Isabella   Peter          5.0
10  Isabella   Ralph          8.0
11      Mary   James          5.0
12      Mary   Peter          4.0
13    Sophia  Robert          5.0
14    Sophia    Eric          4.0
